src/machine/machine_rp2040-rtc.go

```diff
@@ -0,0 +1,256 @@
+		Sec:   int8((rtc_0 & rp.RTC_RTC_0_SEC_Msk) >> rp.RTC_RTC_0_SEC_Pos),
+		Year:  int16((rtc_1 & rp.RTC_RTC_1_YEAR_Msk) >> rp.RTC_RTC_1_YEAR_Pos),
+		Month: int8((rtc_1 & rp.RTC_RTC_1_MONTH_Msk) >> rp.RTC_RTC_1_MONTH_Pos),
+		Day:   int8((rtc_1 & rp.RTC_RTC_1_DAY_Msk) >> rp.RTC_RTC_1_DAY_Pos),
+	}
+
+	return t, nil
+}
+
+// void rtc_set_alarm(datetime_t *t, rtc_callback_t user_callback) {
+func (rtc *rtcType) SetAlarm(t RtcTime, callback func()) {
+
+	rtc.disableInterruptMatch()
+
+	// Only add to setup if it isn't -1
+	// Set the match enable bits for things we care about
+
+	if t.Year >= 0 {
+		rtc.irqSetup0.SetBits(uint32(t.Year) << rp.RTC_SETUP_0_YEAR_Pos)
+		rtc.irqSetup0.SetBits(rp.RTC_IRQ_SETUP_0_YEAR_ENA)
+	}
+
+	if t.Month >= 0 {
+		rtc.irqSetup0.SetBits(uint32(t.Month) << rp.RTC_SETUP_0_MONTH_Pos)
+		rtc.irqSetup0.SetBits(rp.RTC_IRQ_SETUP_0_MONTH_ENA)
+	}
+
+	if t.Day >= 0 {
+		rtc.irqSetup0.SetBits(uint32(t.Day) << rp.RTC_SETUP_0_DAY_Pos)
+		rtc.irqSetup0.SetBits(rp.RTC_IRQ_SETUP_0_DAY_ENA)
+	}
+
+	if t.Dotw >= 0 {
+		rtc.irqSetup1.SetBits(uint32(t.Dotw) << rp.RTC_SETUP_1_DOTW_Pos)
+		rtc.irqSetup1.SetBits(rp.RTC_IRQ_SETUP_1_DOTW_ENA)
+	}
+
+	if t.Hour >= 0 {
+		rtc.irqSetup1.SetBits(uint32(t.Hour) << rp.RTC_SETUP_1_HOUR_Pos)
+		rtc.irqSetup1.SetBits(rp.RTC_IRQ_SETUP_1_HOUR_ENA)
+	}
+
+	if t.Min >= 0 {
+		rtc.irqSetup1.SetBits(uint32(t.Min) << rp.RTC_SETUP_1_MIN_Pos)
+		rtc.irqSetup1.SetBits(rp.RTC_IRQ_SETUP_1_MIN_ENA)
+	}
+
+	if t.Sec >= 0 {
+		rtc.irqSetup1.SetBits(uint32(t.Sec) << rp.RTC_SETUP_1_SEC_Pos)
+		rtc.irqSetup1.SetBits(rp.RTC_IRQ_SETUP_1_SEC_ENA)
+	}
+
+	// Does it repeat? I.e. do we not match on any of the bits
+	rtcAlarmRepeats = t.AlarmRepeats()
+
+	// Store function pointer we can call later
+	rtcCallback = callback
+
+	// Enable the IRQ at the proc
+	interrupt.New(rp.IRQ_RTC_IRQ, rtcHandleInterrupt).Enable()
+	irqSet(rp.IRQ_RTC_IRQ, true)
+
+	// Enable the IRQ at the peri
+	rtc.intE.Set(rp.RTC_INTE_RTC)
+
+	rtc.enableInterruptMatch()
+}
+
+// ---
+
+func (rtc *rtcType) enableInterruptMatch() {
+	// Set matching and wait for it to be enabled
+	rtc.irqSetup0.SetBits(rp.RTC_IRQ_SETUP_0_MATCH_ENA)
+	for !rtc.irqSetup0.HasBits(rp.RTC_IRQ_SETUP_0_MATCH_ACTIVE) {
+	}
+}
+
+func (rtc *rtcType) disableInterruptMatch() {
+	// Disable matching and wait for it to stop being active
+	rtc.irqSetup0.ClearBits(rp.RTC_IRQ_SETUP_0_MATCH_ENA)
+	for rtc.irqSetup0.HasBits(rp.RTC_IRQ_SETUP_0_MATCH_ACTIVE) {
+	}
+}
+
+// ---
+
+func rtcHandleInterrupt(itr interrupt.Interrupt) {
+	// Always disable the alarm to clear the current IRQ.
+	// Even if it is a repeatable alarm, we don't want it to keep firing.
+	// If it matches on a second it can keep firing for that second.
+	RTC.disableInterruptMatch()
+
+	if rtcAlarmRepeats {
+		// If it is a repeatable alarm, re-enable the alarm.
+		RTC.enableInterruptMatch()
+	}
+
+	// Call user callback function
+	if rtcCallback != nil {
+		rtcCallback()
+	}
+}
+
+// ---
+
+type RtcTime struct {
```

Comment:
Should this only be used for setting alarms?   For get/set time, use time.Time instead?

Also, how generic is this - should it be in a top-level `rtc.go`, or is it rp2040 specific?

Reply:
Can't use `time.Time`, unfortunately. See above.

This implementation is RP2040 specific.
We don't have any API for RTC as of yet (nRF family uses it, but does not expose, probably others too).
So the idea was we can try and make it for RP2040 first and generalise once we have at least two families exposing RTC?